wal/src/message_queue_impl/namespace.rs

```diff
@@ -0,0 +1,714 @@
+                namespace: self.namespace.clone(),
+                location,
+            })?
+            .map_or(SequenceNumber::MIN, |table_meta| {
+                table_meta.next_sequence_num - 1
+            }))
+    }
+
+    pub async fn read(
+        &self,
+        ctx: &ReadContext,
+        request: &ReadRequest,
+    ) -> Result<ReadTableIterator<M::ConsumeIterator>> {
+        debug!(
+            "Read table logs in namespace, namespace:{}, ctx:{:?}, request:{:?}",
+            self.namespace, ctx, request
+        );
+
+        let region = self
+            .get_or_open_region(request.location.shard_id as u64)
+            .await
+            .context(ReadWithCause {
+                namespace: self.namespace.clone(),
+                request: request.clone(),
+                msg: "failed while creating iterator",
+            })?;
+
+        let iter = region
+            .scan_table(request.location.table_id, ctx)
+            .await
+            .context(ReadWithCause {
+                namespace: self.namespace.clone(),
+                request: request.clone(),
+                msg: "failed while creating iterator",
+            })?;
+
+        match iter {
+            Some(iter) => ReadTableIterator::new(self.namespace.clone(), iter, request.clone()),
+
+            None => Ok(ReadTableIterator::new_empty()),
+        }
+    }
+
+    pub async fn scan(
+        &self,
+        ctx: &ScanContext,
+        request: &ScanRequest,
+    ) -> Result<ScanRegionIterator<M::ConsumeIterator>> {
+        info!(
+            "Scan region logs in namespace, namespace:{}, ctx:{:?}, request:{:?}",
+            self.namespace, ctx, request
+        );
+
+        let region = self
+            .get_or_open_region(request.region_id)
+            .await
+            .context(ScanWithCause {
+                namespace: self.namespace.clone(),
+                request: request.clone(),
+                msg: "failed while creating iterator",
+            })?;
+
+        let iter = region.scan_region(ctx).await.context(ScanWithCause {
+            namespace: self.namespace.clone(),
+            request: request.clone(),
+            msg: "failed while creating iterator",
+        })?;
+
+        match iter {
+            Some(iter) => Ok(ScanRegionIterator::new(
+                self.namespace.clone(),
+                iter,
+                request.clone(),
+            )),
+
+            None => Ok(ScanRegionIterator::new_empty()),
+        }
+    }
+
+    pub async fn write(
+        &self,
+        ctx: &WriteContext,
+        log_batch: &LogWriteBatch,
+    ) -> Result<SequenceNumber> {
+        debug!(
+            "Write table logs in namespace, namespace:{}, ctx:{:?}, location:{:?}, batch size:{}",
+            self.namespace,
+            ctx,
+            log_batch.location,
+            log_batch.entries.len()
+        );
+
+        let region = self
+            .get_or_open_region(log_batch.location.shard_id as RegionId)
```

Comment:
Actually, I wonder why the `shard_id` is not called `region_id` in the `Location`? `ShardId` should not be the known about the Wal module.

Reply:
I want to do  this together while fixing the bug in obkv implementation.